archive/imo1988_q6.lean

```diff
@@ -0,0 +1,282 @@
+
+/-!
+# IMO1988 Q6 and constant descent Vieta jumping
+
+Question 6 of IMO1988 is somewhat (in)famous. Several expert problem solvers
+could not tackle the question within the given time limit.
+The problem lead to the introduction of a new proof technique,
+so called “Vieta jumping”.
+
+In this file we formalise constant descent Vieta jumping,
+and apply this to prove Q6 of IMO1988.
+To illustrate the technique, we also prove a similar result.
+
+-/
+
+-- open_locale classical
+
+local attribute [instance] classical.prop_decidable
+local attribute [simp] nat.pow_two
+
+/-- Constant descent Vieta jumping is a proof technique
+that can typically be used to show that certain algebraic curves of degree 2
+in the affine plane do not have integral points (or only an explicit set of solutions).
+
+We need a long list of assumptions for this lemma.
+The upside is that it takes care of several annoying edge cases.
+First of all, the user should provide x and y
+that satisfy the predicate H(x,y).
+The functions B(x) and C(x) give the coefficient of the quadratic equation
+y^2 - B(x) * y + C(x) = 0.
+The predicate `base` allows the user to specify a locus of base cases
+that will be proven apart from the descent step `H_desc`.
+This is the end of the input data. What follows are proof obligations.
+In `H_quad` the user must supply a proof that H(x,y) is satisfied
+if and only if the quadratic equation y^2 - B(x) * y + C(x) = 0 is satisfied.
+Next the user should prove that H is symmetric in `H_symm`.
+Then the user should prove H(x,0) and H(x,x) for every x,
+in `H_zero` and `H_diag` respectively.
+Finally, the user should prove the descent step in `H_desc`,
+and the remaining bases cases in `H_base`. -/
+lemma constant_descent_vieta_jumping (x y : ℕ) {claim : Prop} {H : ℕ → ℕ → Prop}
+  (h₀ : H x y) (B : ℕ → ℤ) (C : ℕ → ℤ) (base : ℕ → ℕ → Prop)
+  (H_quad : ∀ {x y}, H x y ↔ (y:ℤ) * y - B x * y + C x = 0) (H_symm : ∀ {x y}, H x y ↔ H y x)
+  (H_zero : ∀ {x}, H x 0 → claim) (H_diag : ∀ {x}, H x x → claim)
+  (H_desc : ∀ {x y}, x < y → x > 0 → ¬base x y → H x y →
+    ∀ y', y' * y' - B x * y' + C x = 0 → y' = B x - y → y' * y = C x → y' ≥ 0 ∧ y' ≤ x)
+  (H_base : ∀ {x y}, H x y → base x y → claim) :
+  claim :=
+begin
+  -- First of all, we may assume that x ≤ y.
+  -- We justify this using H_symm.
+  wlog hxy : x ≤ y, swap, { rw H_symm at h₀, solve_by_elim },
+  -- In fact, we can easily deal with the case x = y.
+  by_cases x_eq_y : x = y, {subst x_eq_y, exact H_diag h₀},
+  -- Hence we may assume that x < y.
+  replace hxy : x < y := lt_of_le_of_ne hxy x_eq_y, clear x_eq_y,
+  -- Consider the upper branch of the hyperbola defined by H.
+  let upper_branch : set (ℕ × ℕ) := {p | H p.1 p.2 ∧ p.1 < p.2},
+  -- Note that the point p = (x,y) lies on the upper branch.
+  let p : ℕ × ℕ := ⟨x,y⟩,
+  have hp : p ∈ upper_branch := ⟨h₀, hxy⟩,
+  -- We also consider the exceptional set of solutions (a,b) that satisfy
+  -- a = 0 or a = b or B a = b or B a = b + a.
```

Comment:
This comment doesn't seem to match the definition--it's missing the "base case". It would be better to say why we doing this, that is, move the "strategy" comment a bit earlier and probably also expand it to contain an outline of the entire rest of the proof.

It seems that the strategy is really: Do descent starting from the given (x, y). The exceptional locus is where we can't keep descending, but our assumptions were chosen to handle those cases.

Reply:
I've tried to improve these comments a bit.